### PR TITLE
feat: add multimodal support to LlamaCppChatGenerator

### LINKED ISSUE #2132: Image support in `LlamaCppChatGenerator`
We want to make this component support multimodal user messages.

Relevant info:
- from llama.cpp python bindings docs (we use this library) https://llama-cpp-python.readthedocs.io/en/latest/#multi-modal-models

### PR DESCRIPTION
**Pull request by ChinmayBansal**
### Related Issues

  - fixes #2132

  ### Proposed Changes:

  This PR adds multimodal (image + text) support to LlamaCppChatGenerator, enabling the component to process
  both text and images in chat messages. The implementation follows established patterns from the
  AnthropicChatGenerator multimodal support (PR #2186).

  **Key Features Added:**
  - Image format validation for supported formats (JPEG, PNG, GIF, WebP)
  - Proper message conversion to LlamaCpp OpenAI-compatible format with base64 data URIs
  - Support for multimodal models through chat handlers and CLIP models
  - Enhanced component initialization with `chat_handler` and `clip_model_path` parameters
  - Role-based image restrictions (images only allowed in user messages)
  - Comprehensive error handling for unsupported formats and edge cases

  **Implementation Details:**
  - Updated `_convert_message_to_llamacpp_format()` function to handle multimodal content while preserving
  order
  - Added multimodal model initialization in `warm_up()` method with Llava15ChatHandler support
  - Enhanced component docstring with detailed usage examples for multimodal scenarios
  - Added proper serialization/deserialization support for new parameters

  ### How did you test it?

  **Unit Tests:**
  - ✅ `test_convert_message_to_llamacpp_format_with_image()` - Tests proper multimodal message conversion
  - ✅ `test_convert_message_to_llamacpp_format_with_unsupported_mime_type()` - Tests error handling for
  unsupported formats
  - ✅ `test_convert_message_to_llamacpp_format_with_none_mime_type()` - Tests edge case with None mime type
  - ✅ `test_convert_message_to_llamacpp_format_image_in_non_user_message()` - Tests role-based restrictions
  - ✅ `test_multimodal_message_processing()` - Tests end-to-end multimodal processing with mocked model

  **Code Quality Verification:**
  - ✅ All linting checks pass: `hatch run fmt`
  - ✅ All type checking passes: `hatch run test:types`
  - ✅ All unit tests pass: `hatch run test:unit`

  **Manual Verification:**
  - Tested multimodal message creation and conversion
  - Verified proper error messages for validation failures
  - Confirmed component initialization with multimodal parameters

  ### Notes for the reviewer

  - The implementation closely follows the patterns established in AnthropicChatGenerator (lines 137-167 in
  anthropic/chat_generator.py)
  - Image validation uses the same error message format as Anthropic for consistency
  - The OpenAI-compatible format with `image_url` structure is required by LlamaCpp for multimodal processing
  - Added comprehensive test coverage that matches and exceeds the patterns used in Anthropic tests
  - All edge cases are properly handled including None mime types and role restrictions

  ### Checklist

  - [x] I have read the [contributors 
  guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the
  [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
  - [x] I have updated the related issue with new insights and changes
  - [x] I added unit tests and updated the docstrings
  - [x] I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for
  my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

### CODE REVIEW

**Comment by anakin87** [comment]
Hello @chimnay, thanks for your work!

In general, I have mixed feeling when I work on this integration.
The main point is that we use `llama-cpp-python` bindings, a project that is not constantly maintained and up to date with llama.cpp features (for example, tool calling lags significantly behind current llama.cpp capabilities). In the long run, I am not sure that depending on this project is the best way to support Haystack users who want to use llama.cpp.

Diving into this PR:
- a core requirement is that the dictionary produced by `to_dict()` is JSON serializable, so we can't directly use the `chat_handler`. I would propose using `chat_handler_name` instead, pointing users to handlers in https://github.com/abetlen/llama-cpp-python/blob/main/llama_cpp/llama_chat_format.py. I think this can then be imported using `importlib`.
- if users specify a `model_clip_path`, they must also provide a `chat_handler_name`
- let's choose a small model (maybe moondream2) and add one integration test
- let's try to remove `type:ignore` wherever possible (sometimes LLMs and coding assistants are not bad at that if you force them for a while :-))

I'm writing this based on reading the llama-cpp-python docs, so I might have missed some details. If that's the case, please let me know.

File: integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py [hunk 1]
```diff
@@ -75,8 +93,33 @@ def _convert_message_to_llamacpp_format(message: ChatMessage) -> ChatCompletionR
...
+            for image in images:
+                if not image.mime_type or image.mime_type not in ["image/jpeg", "image/png", "image/gif", "image/webp"]:
+                    supported_formats = "image/jpeg, image/png, image/gif, image/webp"
+                    msg = (
+                        f"Unsupported image format: {image.mime_type}. "
+                        f"LlamaCpp supports the following formats: {supported_formats}"
+                    )
+                    raise ValueError(msg)
+
+            content_parts = []
```
Comment (anakin87) [inline integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py]: my impression is that we can avoid the mypy error for dict-item
if we declare the type here. Sth like `content_parts:list[dict[str, Any]] = []`

File: integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py [hunk 2]
```diff
@@ -166,6 +240,16 @@ def __init__(
 
         _check_duplicate_tool_names(list(tools or []))
 
+        # Validate multimodal requirements
+        if chat_handler_name is not None:
+            if model_clip_path is None:
+                msg = "model_clip_path must be provided when chat_handler_name is specified for multimodal models"
+                raise ValueError(msg)
+            if chat_handler_name not in SUPPORTED_CHAT_HANDLERS:
```
Comment (anakin87) [inline integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py]: If possible, I would avoid to store `SUPPORTED_CHAT_HANDLERS` to reduce maintenance efforts.
We can try to import the appropriate symbol here in `__init__` and raise an error if that fails. (I would still initialize the handler in `warm_up`).

File: integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py [hunk 3]
```diff
@@ -174,12 +258,30 @@ def __init__(
...
+                handler_path = SUPPORTED_CHAT_HANDLERS[self.chat_handler_name]
+                module_path, class_name = handler_path.rsplit(".", 1)
+                module = importlib.import_module(module_path)
+                chat_handler_class = getattr(module, class_name)
+                chat_handler = chat_handler_class(clip_model_path=self.model_clip_path)
+                self.model_kwargs["chat_handler"] = chat_handler
+            except (ImportError, AttributeError) as e:
+                msg = (
+                    f"Failed to import chat handler '{self.chat_handler_name}' from '{handler_path}'. "
+                    f"Please ensure llama-cpp-python is installed with multimodal support. Error: {e}"
```
Comment (anakin87) [inline integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py]: As I mentioned earlier, I would do the import check in `__init__`.

"Please ensure llama-cpp-python is installed with multimodal support". -> I don't think that there are specific installation options for llama-cpp-python + multimodality. Plus, `llama-cpp-haystack` should take care of installing the required dependencies.

So I would remove this sentence. Please correct me if I am wrong.

File: integrations/llama_cpp/tests/test_chat_generator.py [hunk 4]
```diff
@@ -1096,3 +1241,55 @@ def get_user_detail(name: Annotated[str, "The name of the user"], age: Annotated
         assert "age" in arguments
         assert arguments["name"] == "Jason"
         assert arguments["age"] == 25
+
+    @pytest.mark.integration
+    @pytest.mark.skipif(
+        not os.path.exists("tests/models/moondream2-text-model-f16.gguf")
+        or not os.path.exists("tests/models/moondream2-mmproj-f16.gguf"),
+        reason="Moondream2 models not found. Please download the required models for integration testing.",
+    )
```
Comment (anakin87) [inline integrations/llama_cpp/tests/test_chat_generator.py]: I would download the model on the CI.

For an example, see https://github.com/deepset-ai/haystack-core-integrations/blob/d929ca34d608b428558f421f13f8b4f3f91b31d4/integrations/llama_cpp/tests/test_chat_generator.py#L610

**Review by anakin87 — CHANGES_REQUESTED** [review]
I think we are going in a good direction. I left some comments.

File: integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py (hunk 1 above)
Comment (ChinmayBansal) [reply integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py]: The explicit type declaration was able to  eliminate the type: ingore.

Thanks!

File: integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py (hunk 2 above)
Comment (ChinmayBansal) [reply integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py]: I removed the hard coding supported handlers. I implemented a handler discovery with validation in __init__
I think this is what you were looking for and it would make the code more maintainable.

The new approach discovers all available chathandler classes, generates different name variations to handle different naming patterns, validates imports, and also initializes handlers in warm_up

File: integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py (hunk 3 above)
Comment (ChinmayBansal) [reply integrations/llama_cpp/src/haystack_integrations/components/generators/llama_cpp/chat/chat_generator.py]: There are no separate installation options for multimodal support in llama-cpp-python, and dependencies should be handled by llama-cpp-haystack. The new error message is simplified and more accurate.

File: integrations/llama_cpp/tests/test_chat_generator.py (hunk 4 above)
Comment (ChinmayBansal) [reply integrations/llama_cpp/tests/test_chat_generator.py]: I implemented the model downloading in CI following the suggestion you made.

**Review by anakin87 — APPROVED** [review]
Hey, @ChinmayBansal, thank you again! 

I felt free to simplify some aspects.
I also took the opportunity for using smaller models in the tests, hoping that CI can get faster. I also did another minor adjustments.

I'll merge this PR as soon as tests pass.
